src/crystal/system/win32/delay_load.cr

```diff
@@ -44,6 +45,187 @@ module Crystal::System::DelayLoad
+    end
+
+    ptr = buf
+    {% for i in 0...T.size %}
+      src = args[{{ i }}]
+      while src.value != 0
+        ptr.value = src.value
+        ptr += 1
+        src += 1
+      end
+    {% end %}
+    ptr.value = 0
+
+    {buf, size}
+  end
+
+  # if *str* starts with *prefix*, returns the substring with *prefix* removed,
+  # otherwise returns *str* unmodified
+  private def self.str_lchop(str : LibC::WCHAR*, prefix : LibC::WCHAR*) : LibC::WCHAR*
+    src = str
+
+    while prefix.value != 0
+      return src unless prefix.value == str.value
+      prefix += 1
+      str += 1
+    end
+
+    str
+  end
+
+  # given *str*, a normalized absolute path of *size* UTF-16 code units, returns
+  # its parent directory by replacing the last directory separator with a null
+  # character
+  private def self.dirname(str : LibC::WCHAR*, size : Int32)
+    ptr = str + size - 1
+
+    # C:\foo.exe -> C:
+    # C:\foo\bar.exe -> C:\foo
+    # C:\foo\bar\baz.exe -> C:\foo\bar
+    while ptr != str
+      if ptr.value === '\\'
+        ptr.value = 0
+        return {str, (ptr - str).to_i32!}
+      end
+      ptr -= 1
+    end
+
+    {str, size}
+  end
+
+  # effective returns `::File.dirname(::Process.executable_path).to_utf16`
+  private def self.get_origin! : {LibC::WCHAR*, Int32}
+    buf = LibC.HeapAlloc(LibC.GetProcessHeap, 0, LibC::MAX_PATH &* 2).as(LibC::WCHAR*)
+    len = LibC.GetModuleFileNameW(nil, buf, LibC::MAX_PATH)
+    return dirname(buf, len.to_i32!) unless WinError.value.error_insufficient_buffer?
+
+    buf = LibC.HeapReAlloc(LibC.GetProcessHeap, 0, buf, 65534).as(LibC::WCHAR*)
+    len = LibC.GetModuleFileNameW(nil, buf, 32767)
+    return dirname(buf, len.to_i32!) unless WinError.value.error_insufficient_buffer?
+
+    print_error("FATAL: Failed to get current executable path\n")
+    LibC.ExitProcess(1)
+  end
+
+  # converts *utf8_str* to a UTF-16 string
+  private def self.to_utf16!(utf8_str : LibC::Char*) : LibC::WCHAR*
+    utf16_size = LibC.MultiByteToWideChar(LibC::CP_UTF8, 0, utf8_str, -1, nil, 0)
+    utf16_str = LibC.HeapAlloc(LibC.GetProcessHeap, 0, utf16_size &* 2).as(LibC::WCHAR*)
+    LibC.MultiByteToWideChar(LibC::CP_UTF8, 0, utf8_str, -1, utf16_str, utf16_size)
+    utf16_str
+  end
+
+  # replaces all instances of "$ORIGIN" in *str* with the directory containing
+  # the running executable
+  # if "$ORIGIN" is not found, returns *str* unmodified without allocating
+  # memory
+  private def self.expand_origin!(str : LibC::WCHAR*) : LibC::WCHAR*
+    origin_prefix = UInt16.static_array(0x24, 0x4F, 0x52, 0x49, 0x47, 0x49, 0x4E, 0x00) # "$ORIGIN".to_utf16
+    ptr = str
+    origin = Pointer(LibC::WCHAR).null
+    origin_size = 0
+    output_size = 1
+
+    while ptr.value != 0
+      new_ptr = str_lchop(ptr, origin_prefix.to_unsafe)
+      if new_ptr != ptr
+        origin, origin_size = get_origin! unless origin
+        output_size &+= origin_size
```

Comment:
All the `WCHAR`s in the middle are simply skipped because of `ptr = new_ptr`, and `output_size` did not start with `strlen(str)`, so this is unnecessary